runtime/lua/vim/lsp/inlay_hint.lua

```diff
@@ -4,13 +4,32 @@ local ms = require('vim.lsp.protocol').Methods
 local api = vim.api
 local M = {}
 
----@class (private) vim.lsp.inlay_hint.bufstate
+---@class (private) vim.lsp.inlay_hint.globalstate Global state for inlay hints
+---@field enabled boolean Whether inlay hints are enabled for this scope
+---@type vim.lsp.inlay_hint.globalstate
+local globalstate = {
+  enabled = false,
+}
+
+---@class (private) vim.lsp.inlay_hint.bufstate: vim.lsp.inlay_hint.globalstate Buffer local state for inlay hints
 ---@field version? integer
 ---@field client_hints? table<integer, table<integer, lsp.InlayHint[]>> client_id -> (lnum -> hints)
 ---@field applied table<integer, integer> Last version of hints applied to this line
----@field enabled boolean Whether inlay hints are enabled for this buffer
 ---@type table<integer, vim.lsp.inlay_hint.bufstate>
-local bufstates = {}
+local bufstates = vim.defaulttable(function(_)
+  local bufstate = { applied = {} }
+  setmetatable(bufstate, {
+    __index = globalstate,
+    __newindex = function(state, key, value)
+      if globalstate[key] == value then
+        rawset(state, key, nil)
+      else
+        rawset(state, key, value)
+      end
+    end,
+  })
+  return bufstate
```

Comment:
```suggestion
  return setmetatable({ applied = {} }, {
    __index = globalstate,
    __newindex = function(state, key, value)
      if globalstate[key] == value then
        rawset(state, key, nil)
      else
        rawset(state, key, value)
      end
    end,
  })
```

Reply:
simply ternary
```lua
return setmetatable({applied = {}}, {
    __index = globalstate,
    __newindex = function(state, key, value)
        rawset(state, key, (globalstate[key] == value) and nil or value)
    end
})
```

Reply:
> simply ternary
> 
> ```lua
> return setmetatable({applied = {}}, {
>     __index = globalstate,
>     __newindex = function(state, key, value)
>         rawset(state, key, (globalstate[key] == value) and nil or value)
>     end
> })
> ```

This is incorrect. `nil or value` will always return `value`, the `nil` value will never be chosen

Reply:
I'd say I write this in such a verbose way is I want to emphasize that the `bufstates` is an array of `bufstate`s, and we can only get `bufstate` from it, not anything else. Maybe the effect is not very obvious? I might change it.

Reply:
> I'd say I write this in such a verbose way is I want to emphasize that the `bufstates` is an array of `bufstate`s, and we can only get `bufstate` from it, not anything else. Maybe the effect is not very obvious? I might change it.

I understand, but it's already very clear. It's pretty obvious that an item in a list named `bufstates` will be a single bufstate, so the compact form should be strongly preferred